### PR TITLE
fix(lambda): replace node 8 usages, allow node 10 inline assets

### LINKED ISSUE #4642: [lambda] Support inline lambda code for Node 10.x
CDK doesn't support ZipFile inline lambda code for NODEJS10. It does support it for NodeJs 8, but when the code below uses NodeJs10, CDK returns an error message saying 10 is incompatible.

I have checked this with AWS Support. They're not familiar with the CDK, but doing the same thing in pure CloudFormation works with both 8 and 10

### Reproduction Steps

```
    // This lambda builds the AMI we later will need to start our EC2 instances
    // from. NB that we MUST use InlineCode as the ZipFile format is required
    // for the callback signal from cfnresponse.send to work. So no S3 here.
    const lambdaProvider = new lambda.SingletonFunction(this, 'AmiBuilderLambda', {
      uuid: 'xxxxxxxxxx-xxxxx-xxxx-xxxxxxx-xxxxx',
      runtime: lambda.Runtime.NODEJS_8_10, // MUST NOT be NODEJS 10
      code: new lambda.InlineCode(fs.readFileSync(`${sourcePath}/index.js`, { encoding: 'utf-8' })),
      handler: 'index.handler',
      timeout: cdk.Duration.minutes(15),
      environment: {
          ...
      },
    })
```

### Environment

  - **CLI Version      :1.16.208
  - **Framework Version:1.13.1
  - **OS               :MacOS
  - **Language         :TS

--- 

This is :bug: Bug Report

### LINKED ISSUE #4653: Update deprecated Lambda run-times NODEJS_8_10 to NODEJS_10_X
The Lambda run-time NodeJS 8.10 has been deprecated and you will no longer be able to create new functions with it on January 6, 2020. https://docs.aws.amazon.com/lambda/latest/dg/runtime-support-policy.html

This run-time is in use by a number of Lambda functions that the CDK creates.  For example:
- https://github.com/aws/aws-cdk/blob/592bbc0b7b6baa00248f835899d137beede285a7/packages/%40aws-cdk/aws-dynamodb-global/lib/global-table-coordinator.ts#L19
- https://github.com/aws/aws-cdk/blob/592bbc0b7b6baa00248f835899d137beede285a7/packages/%40aws-cdk/aws-ecr-assets/lib/adopted-repository.ts#L36
- https://github.com/aws/aws-cdk/blob/a9fef66c058a96833b235b4a1690322f0285753c/packages/%40aws-cdk/aws-ses/lib/receipt-rule.ts#L174

It is also used by a lot of integration tests.  These should all be upgraded to use the latest run-time.  AWS has sent out emails to myself and others warning us to upgrade.

### PR DESCRIPTION
**Pull request by nmussy**
* set `supportsInlineCode` to `true` for `Runtime.NODEJS_10_X`
* add `Runtime.NODE_LATEST`, currently aliasing `Runtime.NODEJS_10_X`
* replace all usages of Node 8 and 10 with `Runtime.NODE_LATEST`

Fixes #4653
Fixes #4642

**Before merging**, [aws-cloudformation-coverage-roadmap#80](https://github.com/aws-cloudformation/aws-cloudformation-coverage-roadmap/issues/80) is still "Coming soon", we need to wait for it to be implemented 

---

Alternative: It might be more maintainable to add "latest/stable" runtimes instead, aliasing the preferred version (e.g. `Runtime.NODE_STABLE` pointing to `NODE_10_X`)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-lambda/lib/runtime.ts [hunk 1]
```diff
@@ -34,7 +34,7 @@ export class Runtime {
   public static readonly NODEJS_6_10 =    new Runtime('nodejs6.10',     RuntimeFamily.NODEJS, { supportsInlineCode: true });
   /** @deprecated Use {@link NODEJS_10_X} */
   public static readonly NODEJS_8_10 =    new Runtime('nodejs8.10',     RuntimeFamily.NODEJS, { supportsInlineCode: true });
-  public static readonly NODEJS_10_X =    new Runtime('nodejs10.x',     RuntimeFamily.NODEJS, { supportsInlineCode: false });
+  public static readonly NODEJS_10_X =    new Runtime('nodejs10.x',     RuntimeFamily.NODEJS, { supportsInlineCode: true });
```
Comment (eladb) [inline packages/@aws-cdk/aws-lambda/lib/runtime.ts]: does node10 support inlining? is this new?

Comment (nmussy) [reply packages/@aws-cdk/aws-lambda/lib/runtime.ts]: Seems like it, but I haven't seen any official confirmation yet.

* https://github.com/aws/aws-cdk/issues/4642#issue-511213472
  > I have checked this with AWS Support. They're not familiar with the CDK, but doing the same thing in pure CloudFormation works with both 8 and 10
* https://github.com/aws/aws-cdk/commit/3af61c7e971d0a0ef425eab241c824dd0a9e79d6 built successfully

Comment (jogold) [reply packages/@aws-cdk/aws-lambda/lib/runtime.ts]: The fact that it builds doesn't mean that it works in CF (integ tests are not run).

Just ran an integ test in `eu-west-1` and got the following error message:
> ZipFile can only be used when Runtime is set to either of nodejs, nodejs4.3, nodejs6.10, nodejs8.10, python2.7, python3.6, python3.7.

Comment (nmussy) [reply packages/@aws-cdk/aws-lambda/lib/runtime.ts]: Whoops, thanks @jogold. I thought they were ran.

**Comment by nmussy** [comment]
Could a team member add a "do-not-merge" tag to this PR, waiting for [aws-cloudformation-coverage-roadmap#80](https://github.com/aws-cloudformation/aws-cloudformation-coverage-roadmap/issues/80) to ship?

Thanks!

**Comment by nija-at** [comment]
It looks like CloudFormation now supports inline code for Nodejs10 Lambda functions (tried these out myself).

@nmussy - are you able to move these forward, or do you need help?

**Comment by nmussy** [comment]
@nija-at Great, thanks for letting me know! I'll merge master and see if there are new usages, but this should be good to go

**Comment by nmussy** [comment]
It should be good to go 👍

**Comment by nija-at** [comment]
> I like the NODE_LATEST idea. @nija-at @rix0rrr @RomainMuller what do you guys think?

I have a couple of concerns around `NODE_LATEST` - 

Firstly, does node guarantee that future proofing, i.e., all node8 scripts are valid node10, and so on? If not, we'll be breaking customers (at runtime which is worse) whenever we change `NODE_LATEST`.

Secondly, it is entirely within Lambda's contract and CloudFormation's contract to modify or stop supporting a feature that existed until one major version language to the next. 

A good example is where, for a significant period of time, CloudFormation supported inline code for Node8 but not for Node10. If we had swiched `NODE_LATEST` during this time, we would have broken customers. 

While this was a matter of time, I can other potential situations where such compatibility features  and not available for a while; or upgrade is simply not possible without customers additional or modified attributes. 

How do we handle such scenarios?

File: design/code-asset-metadata.md [hunk 2]
```diff
@@ -75,7 +75,7 @@ const resource = new serverless.CfnFunction(this, 'Func', {
     bucket: asset.s3BucketName,
     key: asset.s3ObjectKey
   },
-  runtime: 'nodejs8.10',
+  runtime: Runtime.NODEJS_LATEST,
```
Comment (nija-at) [inline design/code-asset-metadata.md]: Avoid changes to the `design/` folder unless you're changing the design.

**Comment by nija-at** [comment]
Thinking a bit more on the `NODE_LATEST` feature - this assumes `latest === recommended` which I'm not sure will stay true always.

Given these questions around this and the fact that Node8 is EOL in December (coming around the corner), I suggest we pull out the NODE_LATEST change from here and have that as a separate discussion.

**Comment by nija-at** [comment]
@nmussy - Given that Node8 deprecation is around the corner, I had to separate out the upgrade to Node10 from the `NODE_LATEST` changes.

I didn't have permissions to push on your branch (to update this PR) so I've pulled it out as a separate change https://github.com/aws/aws-cdk/pull/5075/. 
Apologies for the rush on this.

**Comment by nija-at** [comment]
I got the upgrade to Node10 in through another PR - https://github.com/aws/aws-cdk/pull/5075

I'll be closing out on this one. Feel free to open another one if you want to discuss the `NODE_LATEST` feature but before that, do look at the questions raised in this PR.

**Pull request closed without merging**